Add some documentation. (#84)
microsoft/cheriot-rtos: docs/SoftwareCapabilities.md

These APIs combine allocation and sealing, returning both sealed and unsealed capabilities to an object.
The unsealed capability can be used directly, the sealed capability can be passed to other code and unsealed only by calling the allocator's `token_unseal` function with the capability used to allocate the object.

A compartment can use this to generate software-defined capabilities that represent dynamic resources.
For example, a network stack can use it to allocate the state associated with a connection.
The scheduler uses the same mechanism for providing capabilities for cross-thread communication so that, for example, only a holder of the relevant capability can send or receive messages in a message queue.

Static software-defined capabilities
------------------------------------

Dynamic software-defined capabilities, as described in the previous section, are exposed to the allocating compartment and must be passed to other compartments (usually in response to some request).
This presents a bootstrapping problem: what authorises a compartment to request a capability?

> This problem is addressed by providing a mechanism provision software-defined capabilities (managed by one compartment) to another compartment at *build time*.

Reviewer: ```suggestion
This problem is addressed by providing a mechanism for provisioning software-defined capabilities (managed by one compartment) to another compartment at *build time*.
```

### Exporting a sealing type

The `STATIC_SEALING_TYPE` macro defines a new sealing type that the loader can use to seal static objects.
This macro also evaluates to the capability that permits unsealing.

Static sealing capabilities are defined by both their name and the name of the compartment that exports them and so the name that you pick does not need to be unique (though, for documentation purposes, it should not be too generic).

### Creating a sealed value

Static sealed values are created with the `DECLARE_STATIC_SEALED_VALUE` and `DEFINE_STATIC_SEALED_VALUE` macros.
These construct an object with the specified type and contents.

**Note:** These objects may not contain pointers.

### Auditing

Static sealed objects are visible to the linker and set up by the loader.
As such, they are amenable to auditing.
Each static sealing type will have an entry in the exports section of its compartment of the following form:

```json
        { 
          "export_symbol": "__export.sealing_type.static_sealing_inner.SealingType",
          "exported": true,
          "kind": "SealingKey"
        },
```

This example is taken from the test suite.

> It gives the unique symbol name for the sealing key (which includes the compartment name and the key name) and indicates that it is a sealing key.

Reviewer: ```suggestion
It gives the unique symbol name for the sealing key (which includes the compartment name, `static_sealing_inner`, and the key name, `SealingType`) and indicates that it is a sealing key.
```

Reviewer: Mention `exported` too?

Static sealed objects in CHERIoT RTOS
=====================================

CHERIoT is a capability system.
This extends from the hardware (all memory accesses require an authorising capability in a register, provided as an operand to an instruction) up through the software abstractions in the RTOS.
In a capability system, any privileged operation requires an explicit capability to authorise it.

Privilege is a somewhat fluid notion in a system with fine-grained mutual distrust but, in general, any operation that affects state beyond the current compartment is considered privileged.
This includes actions such as:

 - Acknowledging interrupts
 - Allocating heap memory

>  - Establishing network connections

Reviewer 2: More generally, external hardware is "state beyond the current compartment"... include "updating das blinkenlights" and "firing the missiles" in that list?

CHERI provides a hardware mechanism for building software-defined capabilities: sealing.
The sealing mechanism allows a pointer (a CHERI capability) to be made immutable and unusable until it is unsealed using an authorising capability.
The CHERIoT ISA has a limited set of space for sealing types and so these are virtualised with the allocator's [token](../sdk/include/token.h) APIs.
These APIs combine allocation and sealing, returning both sealed and unsealed capabilities to an object.
The unsealed capability can be used directly, the sealed capability can be passed to other code and unsealed only by calling the allocator's `token_unseal` function with the capability used to allocate the object.

A compartment can use this to generate software-defined capabilities that represent dynamic resources.
For example, a network stack can use it to allocate the state associated with a connection.
The scheduler uses the same mechanism for providing capabilities for cross-thread communication so that, for example, only a holder of the relevant capability can send or receive messages in a message queue.

Static software-defined capabilities
------------------------------------

> In addition to the dynamic software-defined capabilities, it is often useful to provision a set of capabilities to a compartment at build time.

Reviewer 2: Maybe draw out the distinction a bit more... something like this, perhaps:
```suggestion
Dynamic software-defined capabilities are (by necessity) exposed to the allocating compartment and must be passed to other compartments (usually in response to some request).
It is often useful to be able to provide software-defined capabilities (managed by one compartment) to another compartment at _build time_, avoiding the need to manage request-response handshake(s) early in initialization.
```

> The simplest case for this is (allocator capabilities)[Allocator.md], which authorise allocation and so are necessary to be able to create any dynamic software-defined capabilities.

Reviewer 2: ```suggestion
The simplest case for this is (allocator capabilities)[Allocator.md], which authorise allocation (and so are necessary to be able to create any dynamic software-defined capabilities!).
```

The scheduler also uses this mechanism for capabilities that allow interaction with [interrupts](Interrupts.md).

> Objects created in this way are allocated outside of a compartment's global region but are accessed only via capabilities provided by the loader.

Reviewer 2: ```suggestion
Objects created in this way are allocated outside of any compartment's global region and are accessed only via capabilities provided by the loader.
The initial contents of these objects are specified during definition, which is often located within a compartment that does not have the rights to unseal the resulting object.
Therefore, these contents, as well as the type of the software-defined capability, are reported in the linker's audit summary and must be validated; see below.
```

Reviewer 2: If that's too many words, it would probably still be good to forward reference the auditing section.

> These capabilities use the allocator's sealing mechanism and so can be unsealed only by the owner of the relevant sealing capability.

Reviewer 2: ```suggestion
These capabilities use the allocator's token mechanism and so can be unsealed only with the relevant authorising capability.
```

For more detail on how to use the static sealing mechanism, see [`compartment-macros.h`](../sdk/include/compartment-macros.h).

### Exporting a sealing type

The `STATIC_SEALING_TYPE` macro defines a new sealing type that the loader can use to seal static objects.
This macro also evaluates to the capability that permits unsealing.

Static sealing capabilities are defined by both their name and the name of the compartment that exports them and so the name that you pick does not need to be unique (though, for documentation purposes, it should not be too generic).

### Creating a sealed value

Static sealed values are created with the `DECLARE_STATIC_SEALED_VALUE` and `DEFINE_STATIC_SEALED_VALUE` macros.
These construct an object with the specified type and contents.

> **Note:** These objects may not contain pointers.

Reviewer 2: What goes wrong if I try?  Is the explosion at compile time, link time, load time, or run time?  If run time, do I just see a 0 tag on the pointer value?

Author: I am not 100% sure.  I believe you will get a capreloc and the loader will say 'this isn't in any compartment' and abort.  We should eventually teach the linker more about where caprelocs are allowed.
